### ISSUE TITLE
Change the training dataset

### ISSUE THREAD

**Issue by 323-w** [Tencent/YOLO-Master#44]
Thank you for your work. I have a few questions I'd like to ask you. 
1. May I ask if this weight file is officially provided? https://huggingface.co/gatilin/YOLO-Master-ckpts-v0/tree/main/YOLO-Master-EsMoE-M
2. I want to train my own detection dataset with only one class. May I ask if I only need to modify the following part of the code: /ultralytics/cfg/models/master/v0_1/det/yolo-master-m.yaml, changing nc: 80 # number of classes to nc:1? I also use my own configuration file mydatasets.yaml. Can I directly use the command yolo detect train data=mydatasets.yaml pretrained=yolo-master-m.pt model=/ultralytics/cfg/models/master/v0_1/det/yolo-master-m.yaml epochs=100 to train? When I train with this command, I find that the training shows relevant loss, but all the test metrics are 0. Could you please tell me what might be the reason for this?

**Comment by 323-w** [comment]
I have another question for you. If I'm doing small object detection, what key parameter values ​​would you recommend?

**Comment by isLinXu** [comment]
Thanks for your interest in YOLO-Master! Let me answer your questions one by one.

## 1. Are the weights official?

Yes. The official pretrained weights are provided in two places:

- **Hugging Face:** [`gatilin/YOLO-Master-ckpts-v0`](https://huggingface.co/gatilin/YOLO-Master-ckpts-v0/tree/main/YOLO-Master-EsMoE-M)
- **GitHub Releases:** [**YOLO-Master-v26.02**](https://github.com/Tencent/YOLO-Master/releases/tag/YOLO-Master-v26.02) — see the **Assets** section at the bottom of the release page, which contains the weight files (incl. `YOLO-Master-EsMoE-M`, deployment formats like `.engine`, etc.) with sha256 checksums.

`YOLO-Master-EsMoE-M` matches the `yolo-master-m` (ES-MoE) config, so it's the right `pretrained=` for `yolo-master-m.yaml`.

## 2. Training on your own single-class dataset

The key thing for one-class detection is to add **`single_cls=True`**:

```
yolo detect train data=mydatasets.yaml model=/ultralytics/cfg/models/master/v0_1/det/yolo-master-m.yaml pretrained=yolo-master-m.pt epochs=100 single_cls=True
```

With `single_cls=True`, Ultralytics treats **all** labels as a single class regardless of their original class index. This is the recommended way to do one-class training, and it also avoids the most common cause of your "metrics = 0" problem (see below). You don't strictly need to hand-edit `nc` in the model YAML — it's overridden from the dataset config / `single_cls` automatically.

**Why loss drops but P/R/mAP are all 0**

The loss is computed against your GT labels, so it falling means training "works"; metrics being 0 means predictions never match GT — almost always a **label/data issue**, not a model issue. Most likely causes, in order:

1. **Class-index mismatch (most common).** Without `single_cls`, if `nc=1` but your label `.txt` files still contain a non-zero class id (e.g. the original COCO id), eval matches nothing → 0. Setting `single_cls=True` collapses everything to class 0 and fixes this. Otherwise, make sure every label id is `0`.
2. **Val split has no matching labels.** Confirm `val:` in `mydatasets.yaml` points to images whose label files exist.
3. **Wrong label format.** Must be `class x_center y_center width height`, all normalized to [0,1].
4. **`nc` vs `len(names)` mismatch** in the dataset YAML.

**Quick diagnosis**

1. Add `single_cls=True` and retrain a few epochs.
2. Check `runs/detect/train/train_batch0.jpg` — are the GT boxes correct?
3. Validate with a low threshold: `yolo detect val model=runs/detect/train/weights/best.pt data=mydatasets.yaml conf=0.001 single_cls=True`.

## 3. Recommended parameters for small object detection

Small objects are exactly where YOLO-Master's **Sparse SAHI** module helps most.

**Use Sparse SAHI (our built-in adaptive slicing).** Instead of feeding the full image at low effective resolution, Sparse SAHI adaptively slices the image so small objects appear larger relative to the receptive field, with ~3–5× speedup over naive SAHI. This is the single most impactful thing for small-object recall. The slicing inference assets and usage are in the release: [**YOLO-Master-v26.02**](https://github.com/Tencent/YOLO-Master/releases/tag/YOLO-Master-v26.02) (see the Assets section).

**Training-side parameters:**
- **`imgsz`** — increase it. `640` → `1024`/`1280` is usually the biggest single win for small objects (at the cost of memory/speed).
- **`scale`** — lower the scale jitter (e.g. `0.5` → `0.2`) so objects aren't shrunk too aggressively during training.
- **`mosaic=1.0`** + **`close_mosaic=10`** — keep mosaic for dense small-object scenes, disabling it only in the final epochs; mild `copy_paste`/`mixup` can boost rare small instances.
- **Stride/heads** — small objects are detected on the high-resolution **P3** (stride 8) head; for extremely tiny objects, adding a **P2** (stride 4) head helps.

**Inference-side parameters:**
- **`conf`** — lower it (e.g. `0.001–0.1` for eval) since small objects produce lower-confidence boxes.
- **`iou` (NMS)** — for dense small objects, our **Cluster-Weighted NMS** avoids over-suppression in crowded scenes.
- **SAHI slice size / overlap** — smaller slices + ~20% overlap recover objects on tile boundaries.

**TL;DR:** `imgsz=1280`, enable Sparse SAHI, lower `conf` at eval, and rely on Cluster-Weighted NMS for dense scenes. Start there and tune slice size/overlap to your object scale.

If anything is still unclear, feel free to share your `mydatasets.yaml`, a sample label `.txt`, and the dataset-scan log header, and we'll dig deeper. 🙌
